[Merged by Bors] - feat(tactic/move_add, test/move_add): a tactic for moving around summands

adomani opened the pull request:
This PR introduces a tactic for moving summands of an expression.  Individual terms can be named, also using pattern-matching, and moved to the beginning or to the end of the sum.

#14228 shows a sample of golfing and usage of `move_add`, and the [diff](https://github.com/leanprover-community/mathlib/compare/aa_sort...adomani_move_add_random_golf) between them.

#14618 contains the doc-string update to `ac_change`: I moved it to a separate PR to save CI cycles.

See `compute_degree` in #14040 for a related tactic to compute `degree`s and `nat_degree`s of polynomials that uses `move_add`.

~~A future PR may add support for sorting also factors of a product, though this may happen after the switch to Lean4.~~

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

- [x] depends on: #14617 for the function `expr.list_summands`.

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

eric-wieser commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,156 @@
+
+/--  Given an expression `ei` and a list `l : list expr`, returns a `tactic expr` that is
+`ei + sum_of_list l`. -/
+meta def build_sum : expr → list expr → tactic expr
```
I would guess this has a one-line implementation with `list.mfold`

adomani replied:
It does!  Thanks for the suggestion!

adomani replied:
I removed this definition and inlined it in the only place where it was used.

eric-wieser commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,156 @@
+`sort_summands_with_weight a wt` returns the list of summands appearing in `a`, sorted using the
+order `<`. -/
+meta def sort_summands_with_weight : expr → (expr → N) → list expr
+| a wt := (get_summands a).qsort (λ e f, wt e < wt f)
```
Could spell this
```suggestion
| a wt := (get_summands a).qsort $ (<) on wt
```
I'd be inclined to move a and wt before the colon too

adomani replied:
I moved the arguments before the colon.

I tried using `on` or `on_fun` or `function.on_fun`, but I think that there is an issue with the fact that `(<)` has Type `N → N → Prop` instead of `N → N → bool`.

eric-wieser replied:
Weird, `merge_sort` takes a Prop relation, but quicksort doesn't

adomani replied:
Doing
```
meta def sort_summands_with_weight (a : expr) (wt : expr → N) : list expr :=
(get_summands a).qsort $ function.on_fun (<) wt
```
says:
```
type mismatch at application
  function.on_fun has_lt.lt
term
  has_lt.lt
has type
  ?m_1 → ?m_1 → Prop : Type ?
but is expected to have type
  ?m_1 → ?m_1 → bool : Sort (max ? 1)
```

adomani replied:
And even with a type annotation:
```
(get_summands a).qsort $ function.on_fun (has_lt.lt : N → N → Prop) wt
```
says
```
type mismatch at application
  function.on_fun has_lt.lt
term
  has_lt.lt
has type
  N → N → Prop : Type u_1
but is expected to have type
  ?m_1 → ?m_1 → bool : Sort (max ? 1)
```

Replacing `Prop` with `bool` complains that the expected type is `Prop`.

adomani replied:
I'm wondering if, instead of a type with `<`, maybe it would be simpler to pass a function `N →N →bool`, use the notation that Eric suggests and then apply this with `has_lt.lt`.

Any opinion?

eric-wieser replied:
Did you miss my comment above about `merge_sort`? That has the nicer interface

adomani replied:
I misunderstood your comment: with `merge_sort`, everything is fine!

eric-wieser commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,156 @@
+| a := do
+  let as := match a.app_fn with
+  | `(coe_fn $ polynomial.monomial %%n) := n.to_nat
+  -- why do I have to remind Lean that this is `with_bot ℕ` and not `option ℕ`?
```
Because you're using `do` notation, which looks for a `monad` instance. I doubt you need `do` here.

adomani replied:
I tried removing `do ... end`, but none of my attempts worked.  Note that this was written by Arthur, so the fact that I failed, does not mean that it is not possible, nor easy!

eric-wieser commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,156 @@
+  n ← get_unused_name,
+  assert n e_eq,
+  -- should we return a meaningful error if this does not manage to prove `e_eq`?
+  reflexivity <|> `[{ simp only [add_comm, add_assoc, add_left_comm], done, }],
```
I think you should be able to use the `ac_change` tactic here

adomani replied:
Ok, but how do I pass the identity to `ac_change` inside the ``[{ ac_change ??? }]`?

eric-wieser replied:
Ah, I guess I mean `ac_refl`, having looked at the implementation of `ac_change`

arthurpaulino replied:
~`ac_refl` loops forever :/~
`ac_refl` is super slow :/

ericrbg replied:
yeah, this tactic will be unusable if it uses `ac_refl`

adomani replied:
Ok, I had left a comment about timings in the example, before knowing about this "feature" of `ac_refl`!

eric-wieser commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,153 @@
+/--  Given an expression `ei` and a list `l : list expr`, returns a `tactic expr` that is
+`ei + sum_of_list l`. -/
+meta def build_sum (ei : expr) (l : list expr) : tactic expr :=
+l.reverse.mfoldr (λ e1 e2, mk_app `has_add.add [e2, e1]) ei
```
Is there an mfoldl?

adomani replied:
`mfoldl` exists and works better, there, thanks!

arthurpaulino committed:
no more dancing junk terms

adomani commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,142 @@
+/--  Given an expression `e` and a weight function `wt : expr → N` to a Type `N` with `<`,
+`sort_summands_with_weight a wt` returns the list of summands appearing in `a`, sorted using the
+order `<`. -/
+meta def sort_summands_with_weight (e : expr) (wt : expr → option N) : list expr :=
```
I am a little bit uneasy about hard-coding that the order is on `option N`.  I'm happy to use a similar ordering in our applications, but `option N` with `compare_fn` is a *special* kind of type with `≤`, so is included as a special case in the def with a general `N`.  I would be happier if we redefined the order that we use on `monomials`, rather than crippling the general ordering possibilities.

adomani commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,142 @@
+
+/--  Given an expression `e` and a weight function `wt : expr → N` to a Type `N` with `<`,
+`sort_summands_with_weight a wt` returns the list of summands appearing in `a`, sorted using the
+order `<`. -/
```
```suggestion
order `≤`. -/
```

adomani commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,142 @@
+end
+
+/--  Given an expression `e` and a weight function `wt : expr → N` to a Type `N` with `<`,
+`sort_summands_with_weight a wt` returns the list of summands appearing in `a`, sorted using the
```
```suggestion
`sort_summands_with_weight a wt` returns the list of summands appearing in `e`, sorted using the
```

arthurpaulino committed:
switching to a custom compare function

arthurpaulino committed:
uncommenting 'sum_sorted_monomials'

adomani committed:
docstrings and capitalization

adomani commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,132 @@
+
+section with_cmp_fn
+
+/--  Given compare function `cmp_fn : expr → expr → bool` and an expression `e`, in an additive
```
```suggestion
/--  Given a compare function `cmp_fn : expr → expr → bool` and an expression `e`, in an
```
Todo: reformulate this so that the "additive commutative semigroup" is more as a motivation.  As is, it seems a requirement.

Possibly "We use this function for expressions in an additive commutative semigroup..."

adomani commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,132 @@
+  e_eq_fmt ← pp e_eq,
+  reflexivity <|>
+    `[{ simp only [add_comm, add_assoc, add_left_comm], done, }] <|>
+    -- `[{ abel, done, }] <|> -- this works too. it's more robust but also a bit slower
```
I think that I prefer the "quick and dirty" argument that we have.  Should it fail often in tests, maybe we can revise this decision.  Does this sound reasonable?

arthurpaulino replied:
Agree

adomani commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,132 @@
+-- `sort_monomials, refl` takes under 300ms
+  sort_monomials,
+  sort_monomials_lhs, -- LHS and RHS agree here
+  sort_monomials_rhs, -- Hmm, both sides change?
```
```suggestion
  sort_monomials_rhs,
```
Now they no longer change, right?

arthurpaulino committed:
allowing custom locations

adomani committed:
doc-strings and disambiguate symbols

adomani committed:
fix?

adomani committed:
doc-strings

arthurpaulino committed:
simplifying 'sort_summands'

adomani committed:
Merge branch 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort

digama0 commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,168 @@
+
+/-- If the target is an equality involving monomials,
+then  `sort_monomials` sorts the summands on either side of the equality. -/
+meta def sort_monomials (locat : parse location) : itactic :=
```
```suggestion
meta def sort_monomials (locat : parse location) : tactic unit :=
```

adomani replied:
Mario, thank you for your suggestion!

Honestly, I am not understanding much of what is going on with the implementation of the tactic and I'm happy that you have looked at it!

adomani committed:
Mario's suggestion

Co-authored-by: Mario Carneiro <di.gama@gmail.com>

adomani committed:
Merge branch 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort

adomani committed:
Merge remote-tracking branch 'origin/master' into aa_sort

arthurpaulino committed:
wiring up for the customized sorting

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort

adomani committed:
rename comp_fn to rel

adomani committed:
more doc-strings

adomani committed:
remove some commented code

adomani committed:
add new sorting function

adomani committed:
add the top list, but probably need to unify with type

adomani committed:
fixes and clean up

adomani committed:
add extra example

adomani committed:
change name to `sort_summands`

adomani committed:
doc-module string

adomani committed:
heuristics

adomani committed:
lint

adomani committed:
more todo

adomani committed:
clean up

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
todo

arthurpaulino commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,186 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Arthur Paulino, Damiano Testa
+-/
+import data.polynomial.degree.definitions
```
This import line should be replaced for something more basic like `tactic.⋯`

adomani replied:
In order to do this, we need to move the heuristics about monomials somewhere else, since otherwise `polynomial.monomial` complains.  I think that it makes sense to do this, but it is also nice to be able to work with a specialized heuristic for monomials.

arthurpaulino commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,186 @@
+  sort_summands [f, g],
+  congr' 3, -- takes care of using assumption `hp`
+  exact X_pow_mul,
+end
```
These should be extracted to a separate test file [here](https://github.com/leanprover-community/mathlib/tree/master/test)

adomani replied:
Done

arthurpaulino commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,186 @@
+  e_eq_fmt ← pp e_eq,
+  reflexivity <|>
+    `[{ simp only [add_comm, add_assoc, add_left_comm], done, }] <|>
+    -- `[{ abel, done, }] <|> -- this works too. it's more robust but also a bit slower
```
What if we use `abel` if the user inputs `sort_summands!`? Or should we wait until we see an example of failure?

adomani replied:
Let's wait and see: since this tactic is only rearranging terms and changing parentheses, I think that invoking `abel` might be overkill.

arthurpaulino reviewed (commented):
As the PR gets readier and readier for review, maybe it's time to move towards these changes

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
reduce imports and create test file

adomani committed:
whitespace

adomani committed:
update todo list

bryangingechen commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,158 @@
+import tactic.core
+import data.polynomial.basic
+
+/-!  # `sort_summands`, a tactic for sorting sums
```
```suggestion
/-!
# `sort_summands`: a tactic for sorting sums
```

bryangingechen commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,158 @@
+    names.mmap $ sort_summands_core ff l,
+    skip
+  end
+
```
There might be some more applicable tags, check the list of tags by expanding "Filter by tag" in the right sidebar [here](https://leanprover-community.github.io/mathlib_docs/tactics.html). Also, since the docstring of `sort_summands` will become the tactic doc on that page, perhaps some of the useful info in the module doc should be copied there.
```suggestion

add_tactic_doc
{ name := "sort_summands",
  category := doc_category.tactic,
  decl_names := [`tactic.interactive.sort_summands],
  tags := ["arithmetic"] }

```

adomani replied:
Thanks Bryan!  I'll look for other applicable tags once I turn on my computer!

adomani replied:
I did not find another suitable tag, so I just left `arithmetic`, that you had suggested.

adomani committed:
Update src/tactic/sort.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

adomani committed:
Update src/tactic/sort.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

adomani committed:
Merge branches 'aa_sort' and 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort

adomani committed:
better doc-string

adomani committed:
move and namespace

adomani committed:
new lemma on initial/final

arthurpaulino commented on src/tactic/sort.lean:
```diff
@@ -0,0 +1,179 @@
+else sort_summands_aux compare_fn t hyp
+
+/-- If the target is an equality involving summands,
+then  `sort_summands` sorts the summands on either side of the equality.
```
It's no longer restricted to equalities. `recurse_on_expr` should make it work with many shapes (=, ≠, <, ≤ etc)

adomani replied:
Amended!

adomani committed:
doc-strings and prepare for lists

adomani committed:
Arthur's fix!

adomani committed:
Arthur's fix!

adomani committed:
doc-strings

adomani committed:
remove unused definition

adomani committed:
Merge branch 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort

adomani committed:
finish merge

arthurpaulino committed:
formatting

adomani committed:
remove old sorting, with top alone

adomani committed:
remove `trace`s, leftover from debugging

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
rearrange sort --> move

adomani committed:
remove `move_add`

adomani committed:
adapt tests

adomani committed:
allow repetitions

adomani committed:
Merge remote-tracking branch 'origin/master' into aa_sort

adomani committed:
spaces and test

adomani committed:
add refl

adomani committed:
try assumption or refl

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
docs and golf

adomani committed:
use `tactic.reflexivity reducible` instead of `reflexivity`

adomani committed:
remove check

eric-wieser commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,224 @@
+
+/--
+Let `rel : expr → expr → bool` be a relation on `expr`s.
+Calling `move_add_with_rel rel [a, ← b, c, ← d]`, recursively looks inside the goal for
```
It would be nice to factor out a non-recursive version, which would be handy in conv mode

adomani replied:
Eric, I like this idea very much: I'll look into it!

I am still trying to get the tactic to behave as I want and I think that I am close now!

adomani replied:
Eric, this means that the tactic should work directly on an expression, assuming that the expression begins being `has_add.add _ _`?

eric-wieser commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,224 @@
+
+Use as `l.sort_with_ends ll rel`.
+ -/
+def list.sort_with_ends {N : Type*} [decidable_eq N] (l : list N) (ll : list N × list N)
```
I think this might be clearer as
```suggestion
def list.sort_with_ends {N : Type*} [decidable_eq N] (l : list N) (init : list N) (tail : list N)
```

adomani replied:
I have redesigned the pre-sorting of the lists.  This should address your comment here and also the one about the `filter` stuff.

eric-wieser commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,224 @@
+  list N :=
+-- the list of initial elements
+let il := ll.1.filter (∈ l) in
+-- the list of final elements
```
Doesn't this behave surprisingly if there are duplicates?

adomani replied:
I think that you are right and I have re-designed the sorting.

adomani committed:
re-design the ordering

adomani committed:
fix example, typo and comment

digama0 commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,245 @@
+
+### Warning:
+* The tactic uses lemmas from `algebra.group.basic`, thus it will not work unless this file is
+  imported
```
Is there a reason you can't just import this file? Or put the tactic in that file.

adomani replied:
No reason, I simply thought that since the file compiled without it, there would have been an `unused imports`-linter!  I added the import.

digama0 commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,245 @@
+* The tactic will discard user-provided terms that do not unify with something in the expression.
+  This means that the tactic will not give an error if it finds no match of the provided terms.
+  The reason for this choice is that we want a single call of `move_add` to move terms across
+  different sums in the same expression.  Here is an example.
```
This rationale only applies if using the tactic in a combinator like `all_goals`, which does not appear to be the case in the examples. If the tactic is handling the iteration itself because it was passed a list of things to do, it can report errors for non-matching terms without affecting the intended behavior.

adomani replied:
Ok, I will work more on understanding how to extract information from the various branches and how to report it and will try to update the tactic and the doc-module.

eric-wieser replied:
To me, this "allow unmatching patterns" behavior feels like it should be hidden behind a flag. (like how `simp` has a flag to allow doing nothing)

adomani replied:
I am trying to catch the presence of unmatching patterns, as per [this discussion](https://leanprover.zulipchat.com/#narrow/stream/239415-metaprogramming-.2F-tactics/topic/catching.20errors.3F/near/280019039)

adomani replied:
The tactic now should report ununified variables and unmodified targets.  It currently treats `at *` as though it was the union of all hypotheses + goal in terms of error reporting, instead of simply making sure that *something* changes (and whathever the correct condition for non-unification should be).

Also, the tests are breaking: I am going to fix them, once the error reporting stabilizes.

I am taking a lunch break now!

adomani replied:
Error reporting for `at *` should be correct now (though I tested it in just a handful of cases).

digama0 commented on src/tactic/move_add.lean:
````diff
@@ -0,0 +1,245 @@
+```
+
+The list of expressions that `move_add` takes is optional and a single expression can be passed
+without parentheses.  Thus `move_add ← f` and `move_add [← f]` mean the same.
````
```suggestion
without brackets.  Thus `move_add ← f` and `move_add [← f]` mean the same.
```

adomani committed:
Update src/tactic/move_add.lean

Co-authored-by: Mario Carneiro <di.gama@gmail.com>

adomani committed:
add import

adomani committed:
Merge branch 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort

adomani committed:
or else!

adomani committed:
remove unused def

adomani committed:
rename find_in2 to find_in

adomani committed:
add message if a hypothesis does not change

adomani committed:
remove commented old code

adomani committed:
more detailed todo

adomani committed:
typo in test

adomani committed:
doc-module

eric-wieser commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,250 @@
+1. we convert a list pairs `bool × pexpr` to a list of pairs `bool × expr`,
+2. we use the extra input `sl : list expr` to perform the unification and sorting step
+   `list.unify_list`,
+3. we jam the  jam the third factor inside the first two.
```
Is this what you meant?
```suggestion
3. we jam the third factor inside the first two.
```

adomani replied:
Thank!  This is what I meant indeed.

eric-wieser commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,250 @@
+  le2 : list expr ← (lp.map (λ e : bool × pexpr, e.2)).mmap to_expr,
+  -- reattach the booleans
+  let tle : list (bool × expr) := (lp.map (λ e : bool × pexpr, e.1)).zip le2,
+  return tle
```
Isn't this something like
```suggestion
lp.mmap $ λ x : bool × pexpr, do
  e ← to_expr x.2,
  return (x.1, e)
```

adomani replied:
Thanks!  I did not think that you could open a `do` block in the middle of... not sure why I did not think you could!

eric-wieser commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,250 @@
+Once we exhausts the elements of `lp`, we return the three lists:
+* first the list of elements of `sl` that came from an element of `lp` whose boolean was `tt`,
+* next the ununified elements of `sl` and
+* finally the elements of `sl` that came from an element of `lp` whose boolean was `ff`.
```
This looks wrong to me - aren't the second and third entries swapped?

adomani replied:
I think that this is correct: if you look at the current version of `list.combined`, you can see that it returns 
```lean
(l1 ++ l3 ++ l2)
```
with the "third" list in the middle.  The third list is the one that is left-over from `sl` after all the unification and removal has taken place.

Note: I am about to push a newer version with a fourth entry, corresponding to the list of unused variables.  While the answer still stands, there will be a fourth list in the product!
EDIT: I am not yet pushing, since trying to report better errors resulted in breaking something else!

eric-wieser replied:
The docstring says "ununified elements of `sl`" appear in the second place. Your comment above says the "third list is the one that is left-over". One of these statements is wrong, and I think it's the docstring.

adomani replied:
Maybe I have looked at this for too long: the third list is the one labeled `l3` and it is placed as the middle sublist of the output.  I am trying to be careful about the third factor in a product and the middle portion of a list that is obtained by concatenating.  I hope that this clarifies.  Otherwise, I will need to take a break, distract myself for a little bit and then come back and read this again!

Thus, the three lists are `(l1, l2, l3)` in a product and I merge them into a single list by concatenating them as `l1 ++ l3 ++ l2`.  Have I made more sense now?

eric-wieser replied:
You're talking about `l1 ++ l3 ++ l2` to describe the "middle" list, but this is the docstring for `list.unify_list`, which does not ever do `l1 ++ l3 ++ l2`; that bit of code is in `list.combined`.

Just look at this line:
```
| [] sl is_unused := return ([], [], sl, is_unused)
```
Clearly the first two lists are the ones unified with the empty list of patterns, and the third one is the one that is not unified. The fact that you later evaluate `[] ++ sl ++ []` isn't relevant to this docstring, because `list.unify_list` isn't the function that does that evaluation.

adomani replied:
I am sorry: you are completely right!  This is the step *before* the one that described in the doc-string.  I will update it with the correct version, which is indeed the one where the lists are "unified with left arrow", next "unified without arrow", finally "left-overs in `sl`".

Thanks!  And sorry for taking so long to realize that what you were saying was correct!

adomani committed:
report unused variables

adomani committed:
Apply suggestions from code review

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

adomani committed:
Merge branch 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort

adomani committed:
lint

adomani committed:
doc-string

adomani committed:
docstring and wild.card

adomani committed:
todo

adomani committed:
better error reporting

adomani committed:
errors for loc.wildcard

adomani committed:
clean up

adomani committed:
tests

adomani committed:
lint

adomani committed:
fix with band?

adomani committed:
remove old defs

adomani committed:
doc-string and remove unused def

adomani commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,313 @@
+The list of expressions that `move_add` takes is optional and a single expression can be passed
+without brackets.  Thus `move_add ← f` and `move_add [← f]` mean the same.
+
+Finally, `move_add` can also be targeted to one or several hypothesis.  If `hp₁, hp₂` are in the
```
```suggestion
Finally, `move_add` can also target one or more hypotheses.  If `hp₁, hp₂` are in the
```

adomani commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,313 @@
+
+##  Implementation notes
+
+The implementation of `move_add` only move the terms specified by the user (and rearranges
```
```suggestion
The implementation of `move_add` only moves the terms specified by the user (and rearranges
```

adomani commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,313 @@
+##  Future work
+
+* Currently, `_`s in user input generate side-goals that Lean should be able to close automatically.
+  Is it possible to avoid actually get Lean to solve these goals right away and not display them?
```
```suggestion
  Is it possible to get Lean to actually solve these goals right away and not display them?
```

adomani commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,313 @@
+  Is it possible to avoid actually get Lean to solve these goals right away and not display them?
+* Using `move_add at *` generates a message mentioning which hypotheses did not change.  It should
+  instead be silent, unless no hypothesis changes.
+* No message is reported if one of the user-provided expressions is not used.
```
```suggestion
```
No longer applies.

adomani committed:
Apply suggestions from code review

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
documentation

adomani committed:
Merge branch 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort

adomani committed:
rename def

adomani committed:
alignment

adomani committed:
reorganize doc-module

eric-wieser commented on src/tactic/move_add.lean:
````diff
@@ -0,0 +1,313 @@
+```
+Finally, `move_add` can also target hypotheses. If `hp` is in the local context,
+`move_add [← f, g] at hp` performs the rearranging at `hp`.
+-/
````
Remember that this docstring is what appears in the tactic documentation page, not the one at the top of the file. It might make sense to shift some text accordingly.

adomani replied:
Thanks, Eric!  I copied most of the doc-module to the doc-string of the tactic.

eric-wieser replied:
I would lean towards not duplicating more than a sentence or two of docstring - instead, just tell the reader in the module docstring to read the docstring of `tactic.interactive.move_add` for more information.

Otherwise there's a large risk of the documentation becoming out of sync and inaccurate.

eric-wieser reviewed (commented):
Can you add a brief comparison to `ac_change` somewhere in the docstring? It might also be worth referencing `move_add` from `ac_change` too.

adomani committed:
doc-string and comparisons

adomani committed:
doc-string

adomani committed:
reference to `move_add` from `ac_change`.

eric-wieser commented on src/tactic/congr.lean:
````diff
@@ -224,6 +224,10 @@ begin
 ```
+
+##  Related tactic: `move_add`
+In the case in which the expression to be changed is a sum of terms, tactic `move_add` can also
````
```suggestion
In the case in which the expression to be changed is a sum of terms, tactic `tactic.interactive.move_add` can also
```
will create a clickable link

eric-wieser commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,389 @@
+higher success rate than `move_add`.  If the goal is an equality of two sums that are simply
+obtained by reparenthesizing and permuting summands, then `move_add [appropriate terms]` can close
+the goal.  Compared to `abel`, `move_add` has the advantage of allowing the user to specify the
+beginning and the end of the final sum, so that from there the user can continue with the proof.
```
These lines should be indented two spaces

adomani replied:
Done!

adomani committed:
deduplicate doc-strings + indentation + link

adomani committed:
Update src/tactic/congr.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

adomani committed:
Merge branch 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort

adomani committed:
Update move_add.lean

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
streamline using failure

adomani committed:
Merge branch 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort

adomani committed:
inline meta def

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
remove comment

adomani committed:
Kyle's suggestion `to_expr tt ff` eliminates unified goals!

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
golf

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
add f

adomani committed:
Add Eric's comment as a todo

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge remote-tracking branch 'origin/master' into aa_sort

robertylewis commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,314 @@
+Used for error management: `un` is the list of user inputs, `bo` is the list encoding which input
+is unused (`tt`) and which input is used (`ff`).
+`list.return_unused` returns the unused user inputs. -/
+def list.return_unused {α : Type*} : list α → list bool → list α
```
If this is going in the `list` namespace I'd prefer it to have a different name, since it will show up in search/autocomplete and there's nothing "unused" about it outside of your application. You could call it `list.filter_by_bool_list` and move it to `data.list.defs`.

robertylewis replied:
Alternatively, this and many of the declarations in this file should be in the `tactic.move_add` namespace.

adomani replied:
Thanks!  I am putting almost all my defs in the `tactic.move_add` namespace and will leave `move_add` alone in the `tactic.interactive` namespace.

I am going through your comments and will push once I am done, but will answer "in real time", to keep track of my progress!

robertylewis commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,314 @@
+def list.return_unused {α : Type*} : list α → list bool → list α
+| un [] := un
+| [] bo := []
+| (u::us) (b::bs) := if b then ([u] ++ (us.return_unused bs)) else (us.return_unused bs)
```
```suggestion
| (u::us) (b::bs) := if b then u::(us.return_unused bs) else us.return_unused bs
```

adomani replied:
Done, but I have removed `list` from the name, so I am not using dot-notation.

robertylewis commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,314 @@
+
+The ununified elements of `sl` get used for error management: they keep track of which user inputs
+are superfluous. -/
+meta def list.unify_list : list (bool × expr) → list expr → list bool →
```
Yeah, this is a pretty specific function related to this tactic. I think it should be in the `move_add` namespace. And `unify_list` isn't a very descriptive name.

adomani replied:
Changed the name to `move_left_or_right`, since this is essentially the use of this definition.

robertylewis commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,314 @@
+  if be.1 then return (ex::l1, l2, l3, is_unused) else return (l1, ex::l2, l3, is_unused)
+
+/--  Given a list of pairs `bool × pexpr`, we convert it to a list of `bool × expr`. -/
+meta def list.convert_to_expr (lp : list (bool × pexpr)) : tactic (list (bool × expr)) :=
```
```suggestion
meta def list.snd_to_expr (lp : list (bool × pexpr)) : tactic (list (bool × expr)) :=
```
but also namespaced. There's also nothing `bool`-specific about this function, the first element of the pair could be of any type.

adomani replied:
name changed and in the more specific `tactic.move_add` namespace.

I also replaced `bool` with an arbitrary type.

robertylewis commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,314 @@
+   `list.unify_list`,
+3. we jam the third factor inside the first two.
+-/
+meta def list.combined (lp : list (bool × pexpr)) (sl : list expr) :
```
Rename

adomani replied:
`combined` --> `final_sorting`

robertylewis commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,314 @@
+setup_tactic_parser
+
+/--  Takes an `expr` and returns a list of its summands. -/
+meta def get_summands : expr → list expr
```
This is in the wrong namespace, it should be `expr.get_summands` and could be moved to `meta/expr.lean`.

adomani replied:
I changed the name to `list_summands`, since it seems closer to the conventions in `meta/expr.lean` and moved it to `meta/expr.lean`.

eric-wieser replied:
Note this is now in #14617. I don't agree that this belongs in `meta/expr.lean`, since it's really a tool for associative operators and there's nowhere near enough infrastructure to talk about associativity in `meta/expr.lean`

robertylewis commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,314 @@
+  (l1, l2, l3, is_unused) ← to_exp.unify_list sl [],
+  return (l1 ++ l3 ++ l2, is_unused)
+
+namespace tactic.interactive
```
You generally shouldn't put things in `tactic.interactive` until the final user-facing tactic. Most of the declarations below should be in the `tactic.move_add` namespace still.

adomani replied:
Thank you!  I did not know and once I push the newer version it should be following these guidelines better!

robertylewis commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,314 @@
+      `[{ simp only [add_comm, add_assoc, add_left_comm], done, }] <|>
+      -- `[{ abel, done, }] <|> -- this works too. it's more robust but also a bit slower
+        fail format!"failed to prove:\n\n{e_eq_fmt}",
+    h ← get_local n,
```
Oh, are you just using `assert` to get a proof term for `e_eq`? `tactic.solve_aux` is the usual solution there. Give it the goal and the `reflexivity <|> simp` script and it will give you a proof term. (And as you do here, make sure to include `done` in the tactic script.)

adomani replied:
As usual, this suggestion works great, thanks!

robertylewis commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,314 @@
+  thyp ← get_local hyp >>= infer_type,
+  is_unused ← recurse_on_expr hyp ll thyp,
+  nthyp ← get_local hyp >>= infer_type,
+  if (thyp = nthyp) then return (tt, is_unused) else return (ff, is_unused)
```
```suggestion
  if thyp = nthyp then return (tt, is_unused) else return (ff, is_unused)
```
unnecessary here and below

robertylewis commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,314 @@
+meta def move_pexpr_list_or_texpr : parser (list (bool × pexpr)) :=
+list_of (move_add_arg 0) <|> list.ret <$> (move_add_arg tac_rbp) <|> return []
+
+/--  Out of a list of `option name`, returns a list of `name`s of target, discarding, if present
```
What does `target` refer to here? Is this just `l.reduce_option.mmap get_local`?

adomani replied:
`target` was my way of referring to a hypothesis in the local context or the goal.

Also, I did not know about the existence of `reduce_option` and your suggestion allowed me to remove this definition and inline it where I used it!  Thanks!

robertylewis commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,314 @@
+
+##  Reporting sub-optimal usage
+
+The tactic never fails (really? TODO), but flags three kinds of unwanted use.
```
I find this suprising as well. It should probably fail if nothing gets changed, or if you refer to variables that aren't present, shouldn't it?

adomani replied:
I think that I prefer the current behaviour: the tactic does not fail, but produces an output saying which hypotheses did not change and which variables are unused.

Changing `trace`s to `fail`s is easy enough, but I find it more friendly if the tactic reports what happened without barking at you!  :smile:

adomani replied:
Also, I've updated the comment about the not failure.

adomani committed:
namespacing and some shifts

adomani committed:
more of Robert's suggestions

adomani committed:
simpler tests

adomani committed:
update following more suggestions

adomani committed:
more tests

adomani committed:
Update src/tactic/move_add.lean

Co-authored-by: Rob Lewis <Rob.y.lewis@gmail.com>

adomani committed:
Merge remote-tracking branch 'origin/master' into aa_sort

adomani committed:
move get_summands to meta.expr list_summands

adomani committed:
remove two more pairs of parens

adomani committed:
remove layered move_add

adomani committed:
test with at h,

adomani committed:
spaces

adomani committed:
remove unneeded open tactic

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
recurse more

adomani committed:
typos

adomani committed:
replace "trace" with "fail"

adomani committed:
update doc-strings to reflect failure

adomani committed:
tests with `success_if_fail`

adomani committed:
clearer doc-string

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
with comments

adomani committed:
without comments

adomani committed:
tests

adomani committed:
test

adomani committed:
doc-string and rename

adomani committed:
whitespace

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
tests

adomani committed:
add `list_summands`

adomani committed:
remove docstring in `tactic.congr`

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
undo change in `meta.expr` for CI cycles

adomani committed:
in the middle

adomani committed:
move_oper

adomani committed:
tests

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
clean up

adomani committed:
error messages

adomani committed:
todo

adomani committed:
lint

adomani committed:
doc-string

adomani committed:
typo

adomani committed:
remove unneeded `open`s

robertylewis commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,415 @@
+  | pes  := fail format!"'{pes}' are unused variables"
+  end,
+  assumption <|> try (tactic.reflexivity reducible)
+  end
```
```suggestion
end
```

robertylewis reviewed (commented):
Again, I'm sorry for taking so long to look at this. It looks great.

Besides sorting out #14617, I'm generally happy to merge. I have one question: you track the unused variables as a list of booleans, then do some scrambling at the end to recover a list of the expressions it corresponds to. Would it make sense to just keep track of a list (or set) of expressions from the beginning? Or am I missing a reason this wouldn't work?

adomani committed:
Update src/tactic/move_add.lean

Co-authored-by: Rob Lewis <Rob.y.lewis@gmail.com>

adomani committed:
remove paragraph

adomani committed:
add Eric's code and a version for factors

adomani committed:
Update src/meta/expr.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

adomani committed:
move `list_binary_operands` to tactic.core

adomani committed:
relate to `list_binary_operands`

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
give possibility of using custom error

adomani committed:
doc-string and rename

adomani committed:
Merge branch 'adomani_list_summands' into aa_sort

adomani committed:
adjust to list_binary_summands

adomani committed:
better example

adomani committed:
factors

adomani committed:
better versions from code review

adomani committed:
Apply suggestions from code review

Co-authored-by: Rob Lewis <Rob.y.lewis@gmail.com>

adomani committed:
rename expr --> tactic

adomani committed:
Merge branch 'adomani_list_summands' of ssh://github.com/leanprover-community/mathlib into adomani_list_summands

adomani committed:
Typo

eric-wieser committed:
add test file

adomani committed:
one more failed test

adomani committed:
Merge branch 'adomani_list_summands' into aa_sort

adomani committed:
use the more stable `tactic.list_binary_operands`

adomani committed:
remove potentially buggy code

adomani committed:
reduce tests

adomani committed:
Merge branch 'adomani_list_summands' into aa_sort

adomani committed:
Update src/tactic/core.lean

Co-authored-by: Rob Lewis <Rob.y.lewis@gmail.com>

adomani committed:
lint

eric-wieser commented on src/tactic/move_add.lean:
````diff
@@ -0,0 +1,397 @@
+Usually, `op` is a binary, associative operation.  E.g.,
+```lean
+#eval trace $ list_head_op ``((+)) tt `(3 / (2 + 4) + 2 * (0 + 2))
+-- [3 / (2 + 4) + 2 * (0 + 2), 2 + 4, 0 + 2]
````
This example isn't a particulary good one, as it doesn't include any terms of the form `a + b + c`

eric-wieser commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,397 @@
+      try $ get_local loc >>= rewrite_hyp h.2,
+      pure is_unused
+    | none     := do
+      try $ rewrite_target h.2,
```
This feels a bit fragile - rather than doing a global rewrite of the expression we found earlier and hoping we match the same one, can we somehow rewrite exactly the occurence that we found like `conv` does? I'm sure there must be some infrastructure for this somewhere...

adomani replied:
Ok, it is indeed fragile, since the term itself may be changed and no longer be available for the rewrite.  Here is an example.  Trying `move_add a * _`

```lean
import tactic.move_add

example {F : Type*} [add_comm_semigroup F] [has_le F] {a b c : F} : (a + b) ≤ (a + b + c) :=
by move_add a
--  ⊢ (b + a) ≤ (b + a + c)
```
whereas I would probably expect the goal to become `(b + a) ≤ (b + c + a)`, moving all occurrences of `a` to the rightmost available position.  I suspect that there are better examples where, once one rewrite takes place, the other rewrite is actually no longer available.  Thus, using the `try` was my way of having a tactic that at least does something rather than failing, even though it may not be exactly what the user intended!

eric-wieser commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,397 @@
+  if F.get_app_fn.const_name = op.get_app_fn.const_name then do
+    Fargs ← list_explicit_args F,
+    ac ← (Fargs ++ [b]).mmap $ list_head_op ff,
+    if bo then return $ [F.mk_app [b]] ++ ac.join
```
Rather than recomputing the application, you could do
```suggestion
| bo F'@(expr.app F b) := do
  op ← to_expr op tt ff,
  if F.get_app_fn.const_name = op.get_app_fn.const_name then do
    Fargs ← list_explicit_args F,
    ac ← (Fargs ++ [b]).mmap $ list_head_op ff,
    if bo then return $ [F'] ++ ac.join
```
Either way this look pretty suspicious to me; I don't think using either `list_explicit_args` or `const_name` is a particularly good idea. I'll have a closer look at this function once the other PR is merged.

adomani replied:
I'm glad that you picked up on this: I was actually thinking of producing a PR with just this function, since I also find it suspicious, after seeing what you said about the previous iterations of `list_summands/operands`.  In fact, in the branch with `compute_degree` I am running into an issue that might be related.  After applying `move_add` (and a bunch of other stuff), I am left with a goal where `norm_num` times out.  If, however I use `R : Type u` with an explicit `universe u`, instead of `R : Type*`, then `norm_num` is instantaneous.  Currently, it is not minimal at all, but I will keep monitoring whether changes to this function make the next one work better, and will report if there are issues.

adomani replied:
As for the change, I tried with this, closer to your solution for `list_operands`: what do you think?

```lean
meta def list_head_op (op : pexpr) : bool → expr → tactic (list expr)
| bo (expr.app F b) := match F with
  | (expr.app g a) := do
    op ← to_expr op tt ff,
    some _ ← try_core (unify op g) | (do Fc ← list_head_op tt F, bc ← list_head_op tt b,
                                      return $ Fc ++ bc),
    Fargs ← list_explicit_args F,
    ac ← (Fargs ++ [b]).mmap $ list_head_op ff,
    if bo then return $ [F.mk_app [b]] ++ ac.join
    else return ac.join
  | _ := do Fc ← list_head_op tt F, bc ← list_head_op tt b,
    return $ Fc ++ bc
  end
| bo (expr.lam _ _ e f) := do
  ec ← list_head_op tt e, fc ← list_head_op tt f,
  return $ ec ++ fc
| bo (expr.pi  _ _ e f) := do
  ec ← list_head_op tt e, fc ← list_head_op tt f,
  return $ ec ++ fc
| bo (expr.mvar  _ _ e) := do
  list_head_op tt e >>= return
| bo (expr.elet _ e f g) := do
  ec ← list_head_op tt e, fc ← list_head_op tt f, gc ← list_head_op tt g,
  return $ ec ++ fc ++ gc
| bo e := return []
```

adomani replied:
I just pushed a version where I merged master and `adomani_list_summands`.

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'adomani_list_summands' into aa_sort

adomani committed:
Update src/tactic/move_add.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort

adomani committed:
better example

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
cleared up

adomani committed:
cosmetic

adomani committed:
a test with forall, exists

adomani committed:
better errors

adomani commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,362 @@
+  (l1, l2, l3, is_unused) ← move_left_or_right l (sl.erase ex) (is_unused.append [ff]),
+  if be.1 then return (ex::l1, l2, l3, is_unused) else return (l1, ex::l2, l3, is_unused)
+
+/--  We combine `snd_to_expr` and `move_left_or_right`, and then some:
```
```suggestion
/--  We adapt `move_left_or_right` to our goal:
```

adomani commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,362 @@
+  if be.1 then return (ex::l1, l2, l3, is_unused) else return (l1, ex::l2, l3, is_unused)
+
+/--  We combine `snd_to_expr` and `move_left_or_right`, and then some:
+1. we convert a list pairs `bool × pexpr` to a list of pairs `bool × expr`,
```
```suggestion
1. we convert a list of pairs `bool × pexpr` to a list of pairs `bool × expr`,
```

adomani commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,362 @@
+This definition is useful to streamline error catching. -/
+meta def with_errors (op : pexpr) (lp : list (bool × pexpr)) (na : option name) :
+  tactic (bool × list bool) :=
+do (thyp, hyploc) ←  -- hyploc is only meaningful in the "is some/else" branch
```
```suggestion
do (thyp, hyploc) ←  -- `hyploc` is only meaningful in the "is some/else" branch
```

adomani commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,362 @@
+  | loc.wildcard := if unch_tgts.band then "'move_op at *' changed nothing" else ""
+  | loc.ns names := let linames := return_unused locas unch_tgts in
+      (if none ∈ return_unused names unch_tgts then "Goal did not change\n" else "") ++
+      (if linames ≠ [] then ("'" ++ to_string linames.reverse ++ "' did not change") else "")
```
`reverse` makes the result print in the "correct" order.  I would have thought that it already was in the correct order, though...

robertylewis commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,362 @@
+  return (expr.macro ma lee, lb.transpose.map list.band)
+| lu e := pure (e, lu)
+
+/-- Parses the user input `ll` to `reorder_oper` at a single location, that could either be
```
What is the input `ll` to `reorder_oper`? This doc is kind of unclear, I don't really understand. Also "replaces" in the description below, in what sense? 

This function could use a more descriptive name too.

adomani replied:
I fixed the doc-string: `ll` should have been `na` and the name is now `reorder_hyp`.

robertylewis reviewed (commented):
@adomani I think this is looking pretty good. Are you happy with the current version?

adomani commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,362 @@
+  return (expr.macro ma lee, lb.transpose.map list.band)
+| lu e := pure (e, lu)
+
+/-- Parses the user input `ll` to `reorder_oper` at a single location, that could either be
```
```suggestion
/-- Parses the user input `na` to `reorder_oper` at a single location, that could either be
```

adomani committed:
Update src/tactic/move_add.lean

adomani committed:
Update src/tactic/move_add.lean

adomani committed:
Update src/tactic/move_add.lean

adomani committed:
Update src/tactic/move_add.lean

adomani commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,362 @@
+`tt` means that the corresponding variable has *not* been unified.
+
+This definition is useful to streamline error catching. -/
+meta def with_errors (op : pexpr) (lp : list (bool × pexpr)) (na : option name) :
```
```suggestion
meta def reorder_hyp (op : pexpr) (lp : list (bool × pexpr)) (na : option name) :
```

adomani commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,362 @@
+locas ← locat.get_locals,
+tg ← target,
+let locas_with_tg := if locat.include_goal then locas ++ [tg] else locas,
+ner ← locas_with_tg.mmap (λ e, with_errors op args e.local_pp_name <|> with_errors op args none),
```
```suggestion
ner ← locas_with_tg.mmap (λ e, reorder_hyp op args e.local_pp_name <|> reorder_hyp op args none),
```

adomani committed:
Update src/tactic/move_add.lean

adomani committed:
Update src/tactic/move_add.lean

adomani commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,362 @@
+
+/-- Parses the user input `na` to `reorder_oper` at a single location, that could either be
+`none` (referring to the goal) or `some name` (referring to hypothesis `name`).  Replaces the
+given location by the rearranged one that `with_errors` receives from `reorder_oper`.
```
```suggestion
`none` (referring to the goal) or `some name` (referring to hypothesis `name`).  Replaces the
given hypothesis/goal with the rearranged one that `reorder_hyp`
receives from `reorder_oper`.
```

adomani committed:
Update src/tactic/move_add.lean

adomani committed:
doc-string

adomani committed:
minor changes

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort

adomani committed:
use alternative

adomani committed:
typo

eric-wieser commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,358 @@
+Once we exhaust the elements of `lp`, we return the three lists:
+* first the list of elements of `sl` that came from an element of `lp` whose boolean was `tt`,
+* next the list of elements of `sl` that came from an element of `lp` whose boolean was `ff`, and
+* finally the ununified elements of `sl`.
```
This function now returns four lists
```suggestion
Once we exhaust the elements of `lp`, we return the four lists:
* `l_tt`: the list of elements of `sl` that came from an element of `lp` whose boolean was `tt`,
* `l_ff`: the list of elements of `sl` that came from an element of `lp` whose boolean was `ff`
* `l_un`: the un-unified elements of `sl`
* `l_m`: a "mask" list of booleans corresponding to the elements of `lp` that were placed in `l_un`
```
Can you rename `l1`, `l2`, `l3` etc in the function to use the same names for clarity?

eric-wieser commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,358 @@
+meta def reorder_oper (op : pexpr) (lp : list (bool × pexpr)) :
+  list bool → expr → tactic (expr × list bool)
+| lu F'@(expr.app F b) := (do op ← to_expr op tt ff,
+    is_given_op op F',
```
A general style comment; I think it's better to wrap this  first line so that it's not missed:
```suggestion
| lu F'@(expr.app F b) := (do
    op ← to_expr op tt ff,
    is_given_op op F',
```
same throughout

eric-wieser commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,358 @@
+    fail format!("the associative/commutative lemmas used do not suffice to prove that " ++
+    "the initial goal equals:\n\n{pre}\n"),
+  if na.is_none then refine ``(eq.mpr %%prf _) else replace_hyp hyploc reordered prf >> skip,
+  return (ff, is_unused)
```
Using `match` feels a bit clearer to me:
```suggestion
meta def reorder_hyp (op : pexpr) (lp : list (bool × pexpr)) (na : option name) :
  tactic (bool × list bool) :=
do
  (thyp, hyploc) ← match na with
    | none := do
        t ← target,
        return (t, none)
    | some na := do
        hl ← get_local na,
        th ← infer_type hl,
        return (th, some hl)
    end,
  (reordered, is_unused) ← reorder_oper op lp (lp.map (λ _, tt)) thyp,
  unify reordered thyp >> return (tt, is_unused) <|> do
  -- the current `do` block takes place where the reordered expression is not equal to the original
  neq ← mk_app `eq [thyp, reordered],
  pre ← pp reordered,
  (_, prf) ← solve_aux neq $
    `[{ simp only [add_comm, add_assoc, add_left_comm], done }] <|>
    `[{ simp only [mul_comm, mul_assoc, mul_left_comm], done }] <|>
    fail format!("the associative/commutative lemmas used do not suffice to prove that " ++
    "the initial goal equals:\n\n{pre}\n"),
  match hyploc with
  | none := refine ``(eq.mpr %%prf _)
  | some hyploc := replace_hyp hyploc reordered prf >> skip
  end,
  return (ff, is_unused)
```
In particular you don't need the `get_or_else` because you already have the de-optioned version

eric-wieser commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,358 @@
+  end,
+guard (str_tgts ++ str_unva = "") <|>
+  fail (if str_tgts.length = 0 then str_unva else str_tgts ++ "\n" ++ str_unva),
+assumption <|> try (tactic.reflexivity reducible)
```
Using `format` instead of `to_string` means the info-view knows how to wrap longer error messages:

```suggestion
meta def move_op (args : parse move_pexpr_list_or_texpr) (locat : parse location) (op : pexpr) :
  tactic unit := do
locas ← locat.get_locals,
tg ← target,
let locas_with_tg := if locat.include_goal then locas ++ [tg] else locas,
ner ← locas_with_tg.mmap (λ e, reorder_hyp op args e.local_pp_name <|> reorder_hyp op args none),
let (unch_tgts, unus_vars) := ner.unzip,
let str_unva := match
  (return_unused args (unus_vars.transpose.map list.band)).map (λ e : bool × pexpr, e.2) with
  | []   := []
  | [pe] := [format!"'{pe}' is an unused variable"]
  | pes  := [format!"'{pes}' are unused variables"]
  end,
let str_tgts := match locat with
  | loc.wildcard := if unch_tgts.band then [format!"nothing changed"] else []
  | loc.ns names := let linames := return_unused locas unch_tgts in
      (if none ∈ return_unused names unch_tgts
        then [format!"Goal did not change"] else []) ++
      (if linames ≠ [] then [format!"'{linames.reverse}' did not change"] else [])
  end,
[] ← pure (str_tgts ++ str_unva) | fail (format.intercalate "\n" (str_tgts ++ str_unva)),
assumption <|> try (tactic.reflexivity reducible)
```
You'll need to remove some trailing `\n`s from the test too.

adomani committed:
Apply Eric's suggestions

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

adomani committed:
formatting

adomani committed:
fix error messages in tests

adomani committed:
rename

adomani committed:
removed two spaces

eric-wieser commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,371 @@
+    (sort_list, is_unused) ← list_binary_operands op F' >>= final_sort lp,
+    sort_all ← sort_list.mmap $ reorder_oper ([lu, is_unused].transpose.map list.band),
+    let (recs, list_unused) := sort_all.unzip,
+    let summed := (recs.drop 1).foldl (λ e f, op.mk_app [e, f]) ((recs.nth 0).get_or_else `(0)),
```
This `` `(0)`` looks suspicious to me; I assume it need to be `` `(1) `` for multiplication? Can you come up with a test that currently fails because of this?

eric-wieser replied:
Ah, I think this is unreachable, because `list_binary_operands` never returns an empty list

adomani replied:
I was going to say this as well.  In fact, I only apply `list_binary_summands` to something that is a sum, so it should never even reach a list with a single element.

adomani replied:
Should I add a comment in the file about the fact that `(0) never really plays a role?

eric-wieser commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,371 @@
+    let summed := (recs.drop 1).foldl (λ e f, op.mk_app [e, f]) ((recs.nth 0).get_or_else `(0)),
+    return (summed, list_unused.transpose.map list.band)) <|>
+    (do [(Fn, unused_F), (bn, unused_b)] ← [F, b].mmap $ reorder_oper lu,
+    return $ (expr.app Fn bn, [unused_F, unused_b, lu].transpose.map list.band))
```
This version with `try_core` is a little clearer than `<|>`, and also makes it possible for error messages to escape:

```suggestion
| lu F'@(expr.app F b) := do
    op ← to_expr op tt ff,
    is_op ← try_core (is_given_op op F'),
    match is_op with
    | some () := do
        (sort_list, is_unused) ← list_binary_operands op F' >>= final_sort lp,
        sort_all ← sort_list.mmap $ reorder_oper ([lu, is_unused].transpose.map list.band),
        let (recs, list_unused) := sort_all.unzip,
        recs_0 :: recs_rest ← pure recs | fail!"internal error: cannot have 0 operands",
        let summed := recs_rest.foldl (λ e f, op.mk_app [e, f]) recs_0,
        return (summed, list_unused.transpose.map list.band)
    | none := do
        [(Fn, unused_F), (bn, unused_b)] ← [F, b].mmap $ reorder_oper lu,
        return $ (expr.app Fn bn, [unused_F, unused_b, lu].transpose.map list.band)
    end
```

adomani committed:
Update src/tactic/move_add.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser reviewed (commented):
Here's an interesting failure case:
```lean
import data.list.of_fn
import tactic.move_add

example (n : ℕ) : list.of_fn (λ i : fin (n + 3), (i : ℕ)) = list.of_fn (λ i : fin (3 + n), i) :=
begin
  move_add [←n],
end
```

Note that `simp_rw [add_comm n 3]` _does_ make progress here; the root cause is that `move_add` does not know about the usual congruence machinery. I suspect that if you went through `tactic.ext_simplify_core` instead of manually recursing then this problem would be resolved.

eric-wieser committed:
Eliminate the `lu` argument to `reorder_oper`.

There is no need to pass this into each level of recursion, and it makes the recursion a little clearer if we don't.

eric-wieser committed:
add a test of metavariables

eric-wieser committed:
replace is_given_op with as_given_op

eric-wieser committed:
use replace_goal to avoid creating metavariable issues

adomani committed:
Merge branch 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort

adomani committed:
update doc-module

eric-wieser commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,378 @@
+* Add operations other than `+` and `*`?  E.g. `∪, ∩, ⊓, ⊔, ...`?
+  Should there be the desire for supporting more operations, it might make sense to extract
+  the `simp [add] <|> simp [mul]` block in `reorder_hyp` to a separate tactic,
+  including all the lemmas used for the rearrangement to work.
```
Can we just use `ac_refl` for this? I understand it's much slower than using the `simp`s, but we can keep those as optimizations for the `+` and `*` cases.

adomani replied:
Is the suggestion to match on the operation, using `simp [whatever]` for add and mul and `ac_refl` otherwise?  I like this idea!  We could also directly support intersections and unions, as it is hardly any more work...

Possibly sups, infs, maxs and mins as well?  I don't know how useful moving those around is.

eric-wieser replied:
I was suggesting the much worse approach of `simp only [add_things] <|> simp only [mul_things] <|> ac_refl`, since I think the latter knows how to do pretty much everything else. The other approach would be to lookup instances of `` `(is_commutative _ %%op)`` and `` `(is_associative _ %%op)`` and then invoke `simp` with lemmas stated in  terms of those typeclasses, which I asume exist somewhere.

adomani replied:
~~I had also thought that we could create a custom typeclass whose fields are op_assoc and op_comm, prove op_left_comm and then simply add instances for all operations that we want to support.~~

Ok, this was a silly idea.

adomani replied:
I just pushed a version that uses `simp [add-lemmas]` for addition, `simp [mul-lemmas]` for multiplication and `ac_refl` otherwise.  I also tried having a general tactic taking the operation as input.

I see two main issues:
1. the operation must be given inside square brackets, since I do not know how to make the parser stop parsing;
2. the `ac_refl` branch only actually works on an equality and does not recurse inside the expression.

adomani committed:
add generic operation

adomani committed:
lint

adomani committed:
typo

adomani committed:
let test

adomani committed:
questions

adomani committed:
golf

adomani committed:
add refl after simp

adomani committed:
add Eric's example to the tests

adomani committed:
typo

adomani committed:
pp errors

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
docs

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'master' into aa_sort

Vierkantor commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,400 @@
+
+Used for error management: `un` is the list of user inputs, `bo` is the list encoding which input
+is unused (`tt`) and which input is used (`ff`).
+`return_unused` returns the unused user inputs. -/
```
```suggestion
`return_unused` returns the unused user inputs.

If `bo` is shorter than `un`, `return_unused` will include the remainder of `un`.
-/
```

Vierkantor commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,400 @@
+  return (l1 ++ l3 ++ l2, is_unused)
+
+/-- `as_given_op op e` unifies the head term of `e` with the binary operation `op`, failing
+if it cannot. -/
```
```suggestion
/-- `as_given_op op e` unifies the head term of `e`, which is a ≥2-argument function application,
with the binary operation `op`, failing if it cannot. -/
```

Vierkantor commented on src/tactic/move_add.lean:
```diff
@@ -0,0 +1,400 @@
+  | `(has_mul.mul) := `[{ simp only [mul_comm, mul_assoc, mul_left_comm]; refl, done }]
+  | _ := ac_refl <|>
+    fail format!("the associative/commutative lemmas used do not suffice to prove that " ++
+      "the initial goal equals:\n\n{pre}\n")
```
I think tactics should make clear when they fail whether the user can do something to fix the issue, or whether this is a bug in the tactic itself. How about this docstring?
```suggestion
    fail format!("the associative/commutative lemmas used do not suffice to prove that " ++
      "the initial goal equals:\n\n{pre}\n" ++
      "Hint: try adding `is_associative` or `is_commutative` instances.\n")
```

Vierkantor commented on src/tactic/move_add.lean:
````diff
@@ -0,0 +1,400 @@
+   (λ (e : ℕ), ∀ (x : ℕ), ∃ (y : ℕ),
+      x * (5 + y + e) + y + 2   = x + e + 2 → x + 2 = 5 + x + y + 2, [ff, ff]) -/
+```
+-/
````
```suggestion

TODO: use `ext_simplify_core` instead of traversing the expression manually
-/
```

Vierkantor reviewed (approved):
Since this PR has been sitting around being nearly ready, I want to go ahead and get this merged.

I agree with @eric-wieser that it's worth trying to use `ext_simplify` instead to get better congruences, but I don't see that we should expect tactics to be able to rewrite the type of a variable `i` from `fin (n + 3)` to `fin (3 + n)` automatically: if `f : fin 3 → ℕ` and `g : (fin a → ℕ) → (fin b → ℕ) → fin (a + b) → ℕ` then `g coe f i` would type-check before rewriting in the type of `i` and fail after rewriting.

So in conclusion, I'm going to follow Rob's point of view and say this is indeed good enough to merge if you are confident you can deal with the bug reports.

bors d+

adomani committed:
Apply Anne's suggestions

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort
